### PR TITLE
[SPARK-14644][ML][PYSPARK] Turn Binary param into a shared param

### PR DESCRIPTION
**Pull request by holdenk**
## What changes were proposed in this pull request?

Change the binary param into a shared param.
## How was this patch tested?

Existing unit tests

### CODE REVIEW

File: python/pyspark/ml/param/shared.py [hunk 1]
```diff
@@ -583,6 +583,31 @@ def getVarianceCol(self):
         return self.getOrDefault(self.varianceCol)
 
 
+class HasBinary(Params):
+    """
+    Mixin for param binary: If True, all non zero results are set to 1. This is useful for discrete probabilistic models that model binary events rather than integer counts. Default False.
```
Comment (rxin) [inline python/pyspark/ml/param/shared.py]: how come this didn't fail our style checker? did we break the pep8 checker?

**Comment by MLnick** [comment]
Hey @holdenk, in https://github.com/apache/spark/pull/12079#issuecomment-209049583 and https://github.com/apache/spark/pull/12308#issuecomment-209039855, we decided to keep them separate for now, since the doc for `binary` in `CountVectorizer` mentions that the binarization occurs after `minTF` is applied. In `HashingTF` this doesn't apply as there is no `minTF` parameter.

So unless we can easily have different doc string, I think it might be best to not separate out `binary` into a shared param yet. We could look at adding `minTF` semantics to `hashingTF`, in which case it would then make a lot of sense to have this as a shared param.

File: python/pyspark/ml/param/shared.py (hunk 1 above)
Comment (holdenk) [inline python/pyspark/ml/param/shared.py]: From toxi.ini we've excluded `cloudpickle.py`,`heapq3.py`, and `shared.py`, we've also done similar exclusions on the Scala side for the auto generated params.

**Comment by holdenk** [comment]
So I thought the unified doc string of:

> If true, all non zero results are set to 1. This is useful for discrete probabilistic models that model binary events rather than integer counts.  Default False.

might make sense in the two cases where we are using it (since `minTF` is described as a filter). If you think this isn't clear enough with the `minTF` case on `HashingTF` I'm happy to close it for now and revisit if we add `minTF` to `CountVectorizer`.

**Comment by MLnick** [comment]
ah - sorry I missed that. Ok, yeah this makes sense if we unify the doc string appropriately. Perhaps something like `If true, all non-zero counts (after any filters are applied) ...`?

**Comment by holdenk** [comment]
Sure that sounds like a clearer docstring :)

**Comment by jkbradley** [comment]
I prefer to keep them separate.  There isn't much benefit to combining 2 instances, and the doc becomes a bit less clear.

**Comment by MLnick** [comment]
Fair enough. We can revisit if more estimators use the param, or if we add
minTF to HashingTF.

**Comment by holdenk** [comment]
Sounds good if people prefer to keep them separate I'll close this out.

**Pull request closed without merging**